src/runtime_src/core/common/drv/kds_core.c

```diff
@@ -130,13 +163,125 @@ int kds_init_client(struct kds_client *client)
 	return 0;
 }
 
-void kds_fini_client(struct kds_client *client)
+void kds_fini_client(struct kds_sched *kds, struct kds_client *client)
 {
+	struct kds_ctx_info info;
+	struct kds_ctrl *ctrl;
+
 #if PRE_ALLOC
 	vfree(client->xcmds);
 	vfree(client->infos);
 #endif
-	return;
+	ctrl = client->ctrl[KDS_CU];
+	info.flags = CU_CTX_OP_FINI;
+	if (ctrl)
+		ctrl->control_ctx(ctrl, client, &info);
+
+	put_pid(client->pid);
+	mutex_destroy(&client->lock);
+
+	mutex_lock(&kds->lock);
+	list_del(&client->link);
+	kds->num_client--;
+	mutex_unlock(&kds->lock);
+}
+
+int kds_add_context(struct kds_client *client, struct kds_ctx_info *info)
+{
+	u32 cu_idx = info->cu_idx;
+	bool shared = (info->flags != CU_CTX_EXCLUSIVE);
+	struct kds_ctrl *ctrl = client->ctrl[KDS_CU];
+
+	BUG_ON(!mutex_is_locked(&client->lock));
+
+	/* TODO: In lagcy KDS, there is a concept of implicit CUs.
+	 * It looks like that part is related to cdma. But it use the same
+	 * cu bit map and it relies on to how user open context.
+	 * Let's consider that kind of CUs later.
+	 */
+	if (cu_idx == CU_CTX_VIRT_CU) {
+		if (!shared) {
+			kds_err(client, "Only allow share virtual CU");
+			return -EINVAL;
+		}
+		++client->virt_cu_ref;
+	} else {
+		if (!ctrl)
+			return -ENODEV;
+		info->flags &= ~CU_CTX_OP_MASK;
+		info->flags |= CU_CTX_OP_ADD;
+		if (ctrl->control_ctx(ctrl, client, info))
+			return -EINVAL;
+	}
+
+	++client->num_ctx;
+	kds_info(client, "Client pid(%d) add context CU(0x%x) shared(%s)",
+		 pid_nr(client->pid), cu_idx, shared? "true" : "false");
+	return 0;
+}
+
+int kds_del_context(struct kds_client *client, struct kds_ctx_info *info)
+{
+	u32 cu_idx = info->cu_idx;
+	struct kds_ctrl *ctrl = client->ctrl[KDS_CU];
+
+	BUG_ON(!mutex_is_locked(&client->lock));
+
+	if (cu_idx == CU_CTX_VIRT_CU) {
+		if (!client->virt_cu_ref) {
+			kds_err(client, "No opening virtual CU");
+			return -EINVAL;
+		}
+		--client->virt_cu_ref;
+	} else {
+		if (!ctrl)
+			return -ENODEV;
+		info->flags &= ~CU_CTX_OP_MASK;
+		info->flags |= CU_CTX_OP_DEL;
+		if (ctrl->control_ctx(ctrl, client, info))
+			return -EINVAL;
+	}
+
+	--client->num_ctx;
+	kds_info(client, "Client pid(%d) del context CU(0x%x)",
+		 pid_nr(client->pid), cu_idx);
+	return 0;
+}
+
+u32 kds_live_clients(struct kds_sched *kds, pid_t **plist)
+{
+	const struct list_head *ptr;
+	struct kds_client *client;
+	pid_t *pl = NULL;
+	u32 count = 0;
+	u32 i = 0;
+
+	mutex_lock(&kds->lock);
+	/* Find out number of active client */
+	list_for_each(ptr, &kds->clients) {
+		client = list_entry(ptr, struct kds_client, link);
+		if (client->num_ctx > 0)
+			count++;
+	}
+	if (count == 0 || plist == NULL)
+		goto out;
```

Comment:
do we need to release lock before bailing out?

Reply:
fixed